ext/opentelemetry-ext-pymemcache/src/opentelemetry/ext/pymemcache/util.py

```diff
@@ -0,0 +1,88 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Some utils used by the pymemcache integration
+"""
+
+
+# Network attribute semantic convention here:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-network-connection-attributes
+_HOST = "net.peer.name"
+_PORT = "net.peer.port"
+# Database semantic conventions here:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md
+_DB = "db.type"
+_URL = "db.url"
+
+
+def _get_address_attributes(instance):
+    """Attempt to get host and port from Client instance."""
+    address_attributes = {}
+    address_attributes[_DB] = "memcached"
+
+    # client.base.Client contains server attribute which is either a host/port tuple, or unix socket path string
+    # https://github.com/pinterest/pymemcache/blob/f02ddf73a28c09256589b8afbb3ee50f1171cac7/pymemcache/client/base.py#L228
+    try:
+        if hasattr(instance, "server"):
+            if isinstance(instance.server, tuple):
+                host, port = instance.server
+                address_attributes[_HOST] = host
+                address_attributes[_PORT] = port
+                address_attributes[_URL] = "memcached://{}:{}".format(
+                    host, port
+                )
+            elif isinstance(instance.server, str):
+                address_attributes[_URL] = "memcached://{}".format(
+                    instance.server
+                )
+
+    except Exception:  # pylint: disable=broad-except
+        pass
+
+    return address_attributes
+
+
+def _get_query_string(args):
+
+    """Return the query values given the arguments to a pymemcache command.
+
+    If there are multiple query values, they are joined together
+    space-separated.
+    """
+    keys = ""
+
+    # shortcut if no args
+    if not args:
+        return keys
+
+    # pull out the first arg which will contain any key
+    arg = args[0]
+
+    # if we get a dict, convert to list of keys
+    if isinstance(arg, dict):
+        arg = list(arg)
+
+    if isinstance(arg, str):
+        keys = arg
+    elif isinstance(arg, bytes):
+        keys = arg.decode()
+    elif isinstance(arg, list) and len(arg) >= 1:
+        if isinstance(arg[0], str):
```

Comment:
are we guaranteed that every argument in the argument list is a string if the first one is? not familiar with pymemcache, so maybe it is guarenteed

Reply:
Not a pymemcache expert but yes I believe that's the case, first arg is either:

- keys, a list or dict of key strings, like here: `https://github.com/pinterest/pymemcache/blob/f02ddf73a28c09256589b8afbb3ee50f1171cac7/pymemcache/client/base.py#L503`
- or a single key, a string: `https://github.com/pinterest/pymemcache/blob/f02ddf73a28c09256589b8afbb3ee50f1171cac7/pymemcache/client/base.py#L324`

Also, just to clarify, we're only grabbing the keys of the query, not the full command